trustfall_core/src/frontend/error.rs

```diff
@@ -109,6 +109,12 @@ pub enum FrontendError {
     #[error("The query failed to validate against the schema: {0}")]
     ValidationError(#[from] ValidationError),
 
+    #[error(
+        "The field \"{0}\" is a property but the query uses @fold.\
+     Only edges can be used with @fold."
+    )]
+    FoldOnProperty(String),
+
```

Comment:
A few questions here. I genuinely don't know the answer to these off the top of my head, but we probably want to consider them so as to make consistent and intuitive choices:
- What error does the frontend raise for other edge-only directives (`@optional` / `@recurse`) is used on a property? (Not sure if we already have test cases for these, but we should add them if we don't!)
- Would it make sense to use a more general variant for all such directives that are invalid on properties, like a `UnsupportedDirectiveOnProperty`? (Name subject to workshopping, open to ideas.)

Also, a couple of nits:
- We are not guaranteeing enum variant order, so let's try to group this variant near other variants for related errors instead.
- The error message is a bit unclear: for example, it only says that `@fold` was used by never says the actual problem — that it was used by applying it _to a property_. The existing error messages here are also not good, but good diagnostics are super important so we should try to start raising the quality bar if at all possible.

Reply:
> What error does the frontend raise for other edge-only directives (`@optional` / `@recurse`) is used on a property? (Not sure if we already have test cases for these, but we should add them if we don't!)

Unless the playground is outdated and validations were added after, those do not raise any errors. So it feels like this change should include those directives as well and use naming like you proposed (`UnsupportedDirectiveOnProperty`)

> We are not guaranteeing enum variant order, so let's try to group this variant near other variants for related errors instead.
> The error message is a bit unclear: for example, it only says that `@fold` was used by never says the actual problem — that it was used by applying it to a property. The existing error messages here are also not good, but good diagnostics are super important so we should try to start raising the quality bar if at all possible.

I will send out a version with a little better wording, but I will sleep on it to see if I can improve it.

Reply:
I don't think any validations have been added in a while. Oops! Great catch, thanks for finding this!